Test flakes: TestHandleConnection fails during cleanup

It appears our test func returns before things are fully shut down, and when the runtime attempts to clean up the `t.TempDir()` something is still writing to it.

I think this happens in several places, but here's a log of a recent example.

```
FAIL: github.com/gravitational/teleport/lib/srv/app
852	FAIL: github.com/gravitational/teleport/lib/srv/app.TestHandleConnection
853	===================================================
854	OUTPUT github.com/gravitational/teleport/lib/srv/app.TestHandleConnection
855	===================================================
856	=== RUN   TestHandleConnection
857	2021-11-05T18:00:26Z DEBU [KEYGEN]    SSH cert authority started with no keys pre-compute. native/native.go:106
858	2021-11-05T18:00:27Z DEBU             Starting watch. resource-kind:lock retry:Linear(attempt=0, duration=0s) services/watcher.go:177
859	2021-11-05T18:00:27Z DEBU [BUFFER]    Add Watcher(name=auth, prefixes=/locks, capacity=1024, size=0). backend/buffer.go:312
860	2021-11-05T18:00:27Z DEBU [KEYGEN]    Generated SSH host certificate for role Auth with principals: [test-tls-server.root.example.com test-tls-server localhost 127.0.0.1 ::1]. native/native.go:231
861	2021-11-05T18:00:27Z INFO [CA]        Generating TLS certificate {0xc0007c9200 0xc000798b00 1.3.9999.1.7=#1310726f6f742e6578616d706c652e636f6d,CN=test-tls-server.root.example.com,O=Auth,POSTALCODE=null,STREET= 1994-04-02 00:00:00 +0000 UTC [*.teleport.cluster.local teleport.cluster.local] [] 5 []}. common_name:test-tls-server.root.example.com dns_names:[*.teleport.cluster.local teleport.cluster.local] locality:[] not_after:1994-04-02 00:00:00 +0000 UTC org:[Auth] org_unit:[] tlsca/ca.go:650
862	2021-11-05T18:00:27Z DEBU [AUTH]      GRPC(SERVER): keep alive 1m0s count: 3. auth/grpcserver.go:3493
863	2021-11-05T18:00:27Z DEBU [KEYGEN]    Generated SSH host certificate for role App with principals: [491c11f0-63b1-4804-9fc1-1930f7cbb284.root.example.com 491c11f0-63b1-4804-9fc1-1930f7cbb284 localhost 127.0.0.1 ::1]. native/native.go:231
864	2021-11-05T18:00:27Z INFO [CA]        Generating TLS certificate {0xc0007c9200 0xc000869ed0 1.3.9999.1.7=#1310726f6f742e6578616d706c652e636f6d,CN=491c11f0-63b1-4804-9fc1-1930f7cbb284.root.example.com,O=App,POSTALCODE=null,STREET= 1994-04-02 00:00:00 +0000 UTC [*.teleport.cluster.local teleport.cluster.local] [] 5 []}. common_name:491c11f0-63b1-4804-9fc1-1930f7cbb284.root.example.com dns_names:[*.teleport.cluster.local teleport.cluster.local] locality:[] not_after:1994-04-02 00:00:00 +0000 UTC org:[App] org_unit:[] tlsca/ca.go:650
865	2021-11-05T18:00:27Z DEBU [KEYGEN]    Generated SSH host certificate for role Auth with principals: [491c11f0-63b1-4804-9fc1-1930f7cbb284.root.example.com 491c11f0-63b1-4804-9fc1-1930f7cbb284 localhost 127.0.0.1 ::1]. native/native.go:231
866	2021-11-05T18:00:27Z INFO [CA]        Generating TLS certificate {0xc0007c9200 0xc0008752b0 1.3.9999.1.7=#1310726f6f742e6578616d706c652e636f6d,CN=491c11f0-63b1-4804-9fc1-1930f7cbb284.root.example.com,O=Auth,POSTALCODE=null,STREET= 1994-04-02 00:00:00 +0000 UTC [*.teleport.cluster.local teleport.cluster.local] [] 5 []}. common_name:491c11f0-63b1-4804-9fc1-1930f7cbb284.root.example.com dns_names:[*.teleport.cluster.local teleport.cluster.local] locality:[] not_after:1994-04-02 00:00:00 +0000 UTC org:[Auth] org_unit:[] tlsca/ca.go:650
867	2021-11-05T18:00:27Z DEBU [AUTH]      ClientCertPool -> cert(root.example.com issued by root.example.com:243004124792885478439102525550816744087) auth/middleware.go:619
868	2021-11-05T18:00:27Z DEBU [AUTH]      ClientCertPool -> cert(root.example.com issued by root.example.com:65799196625339907368122223959506757244) auth/middleware.go:619
869	2021-11-05T18:00:27Z DEBU [AUTH]      Server certificate cert(test-tls-server.root.example.com issued by root.example.com:243004124792885478439102525550816744087). auth/middleware.go:300
870	2021-11-05T18:00:27Z DEBU [KEYGEN]    generated user key for [foo] with expiry on (449888400) 1984-04-04 01:00:00 +0000 UTC native/native.go:256
871	2021-11-05T18:00:27Z DEBU [AUTH]      Failed setting default kubernetes cluster for user login (user did not provide a cluster); leaving KubernetesCluster extension in the TLS certificate empty auth/auth.go:938
872	2021-11-05T18:00:27Z INFO [CA]        Generating TLS certificate {0xc0007c9200 0xc00020dfc0 1.3.9999.1.7=#1310726f6f742e6578616d706c652e636f6d,1.3.9999.1.12=#1308726561646f6e6c79,1.3.9999.1.5=#1310726f6f742e6578616d706c652e636f6d,1.3.9999.1.6=#130f666f6f2e6578616d706c652e636f6d,1.3.9999.1.4=#132464643835653665322d643236392d346463612d616235382d363263303465306532316435,CN=foo,OU=usage:apps,O=user:foo,POSTALCODE={\"logins\":[\"535e95bd-5013-4c88-8026-60bd7b91fe05\"]},STREET=root.example.com,L=foo 1984-04-04 01:00:00 +0000 UTC [] [] 5 []}. common_name:foo dns_names:[] locality:[foo] not_after:1984-04-04 01:00:00 +0000 UTC org:[user:foo] org_unit:[usage:apps] tlsca/ca.go:650
873	2021-11-05T18:00:27Z DEBU [KEYGEN]    generated user key for [foo] with expiry on (449888400) 1984-04-04 01:00:00 +0000 UTC native/native.go:256
874	2021-11-05T18:00:27Z DEBU [AUTH]      Failed setting default kubernetes cluster for user login (user did not provide a cluster); leaving KubernetesCluster extension in the TLS certificate empty auth/auth.go:938
875	2021-11-05T18:00:27Z INFO [CA]        Generating TLS certificate {0xc0007c9200 0xc0008682a0 1.3.9999.1.7=#1310726f6f742e6578616d706c652e636f6d,1.3.9999.1.12=#1308726561646f6e6c79,1.3.9999.1.11=#1308726561646f6e6c79,1.3.9999.1.5=#1310726f6f742e6578616d706c652e636f6d,1.3.9999.1.6=#130f6177732e6578616d706c652e636f6d,1.3.9999.1.4=#132430383661376139362d373166352d346639312d396137622d346333343831346237626136,CN=foo,OU=usage:apps,O=user:foo,POSTALCODE={\"logins\":[\"56036c79-e8de-4135-819c-04958ef9bf30\"]},STREET=root.example.com,L=foo 1984-04-04 01:00:00 +0000 UTC [] [] 5 []}. common_name:foo dns_names:[] locality:[foo] not_after:1984-04-04 01:00:00 +0000 UTC org:[user:foo] org_unit:[usage:apps] tlsca/ca.go:650
876	2021-11-05T18:00:27Z DEBU             Starting watch. resource-kind:lock retry:Linear(attempt=0, duration=0s) services/watcher.go:177
877	2021-11-05T18:00:27Z DEBU [BUFFER]    Add Watcher(name=491c11f0-63b1-4804-9fc1-1930f7cbb284.root.example.com, prefixes=/locks, capacity=1024, size=0). backend/buffer.go:312
878	2021-11-05T18:00:27Z DEBU [AUTH]      ClientCertPool -> cert(root.example.com issued by root.example.com:243004124792885478439102525550816744087) auth/middleware.go:619
879	2021-11-05T18:00:27Z DEBU [AUTH]      ClientCertPool -> cert(root.example.com issued by root.example.com:65799196625339907368122223959506757244) auth/middleware.go:619
880	2021-11-05T18:00:27Z DEBU [AUTH]      Server certificate cert(test-tls-server.root.example.com issued by root.example.com:243004124792885478439102525550816744087). auth/middleware.go:300
881	2021-11-05T18:00:27Z DEBU [APP:SERVI] Starting App heartbeat with announce period: 1m0s, keep-alive period 5m13.112246955s, poll period: 5s srv/heartbeat.go:152
882	2021-11-05T18:00:27Z DEBU [APP:SERVI] Started App(Name=foo, PublicAddr=foo.example.com, Labels=map[bar:baz qux:]). app/server.go:277
883	2021-11-05T18:00:27Z DEBU [APP:SERVI] Starting App heartbeat with announce period: 1m0s, keep-alive period 5m24.666323651s, poll period: 5s srv/heartbeat.go:152
884	2021-11-05T18:00:27Z DEBU [APP:SERVI] Started App(Name=awsconsole, PublicAddr=aws.example.com, Labels=map[bar:baz]). app/server.go:277
885	2021-11-05T18:00:27Z DEBU [APP:SERVI] Not initializing application resource watcher. app/watcher.go:66
886	2021-11-05T18:00:27Z DEBU [APP:SERVI] Forcing heartbeat for "foo". app/server.go:536
887	2021-11-05T18:00:27Z DEBU [APP:SERVI] Forcing heartbeat for "awsconsole". app/server.go:536
888	2021-11-05T18:00:27Z DEBU [APP:SERVI] Asked check out of cycle srv/heartbeat.go:276
889	2021-11-05T18:00:27Z DEBU [APP:SERVI] Asked check out of cycle srv/heartbeat.go:276
890	2021-11-05T18:00:27Z DEBU [AUTH:GRPC] Got heartbeat connection from 491c11f0-63b1-4804-9fc1-1930f7cbb284.root.example.com. auth/grpcserver.go:129
891	2021-11-05T18:00:27Z DEBU [AUTH:GRPC] Got heartbeat connection from 491c11f0-63b1-4804-9fc1-1930f7cbb284.root.example.com. auth/grpcserver.go:129
892	2021-11-05T18:00:27Z DEBU [AUTH]      ClientCertPool -> cert(root.example.com issued by root.example.com:243004124792885478439102525550816744087) auth/middleware.go:619
893	2021-11-05T18:00:27Z DEBU [AUTH]      ClientCertPool -> cert(root.example.com issued by root.example.com:65799196625339907368122223959506757244) auth/middleware.go:619
894	2021-11-05T18:00:27Z DEBU [RBAC]      Access to app "foo" granted, allow rule in role "user:foo" matched. services/role.go:1650
895	2021-11-05T18:00:27Z DEBU [APP:SERVI] Using async streamer for session fec7204e-81ca-4ec0-90a6-75e3acb26b23. app/session.go:184
896	2021-11-05T18:00:27Z INFO             Round trip: GET /, code: 200, duration: 1.40115ms tls:version: 304, tls:resume:false, tls:csuite:1301, tls:server:teleport.cluster.local forward/fwd.go:185
897	2021-11-05T18:00:27Z DEBU [APP:SERVI] Heartbeat exited. srv/heartbeat.go:278
898	2021-11-05T18:00:27Z DEBU [AUTH:GRPC] Failed to receive heartbeat: rpc error: code = Canceled desc = context canceled auth/grpcserver.go:139
899	2021-11-05T18:00:27Z DEBU [APP:SERVI] Stopped app "foo". app/server.go:292
900	2021-11-05T18:00:27Z DEBU [APP:SERVI] Heartbeat exited. srv/heartbeat.go:278
901	2021-11-05T18:00:27Z DEBU [AUTH:GRPC] Failed to receive heartbeat: rpc error: code = Canceled desc = context canceled auth/grpcserver.go:139
902	2021-11-05T18:00:27Z DEBU [APP:SERVI] Stopped app "awsconsole". app/server.go:292
903	2021-11-05T18:00:27Z WARN [APP:SERVI] Failed to handle connection: http: Server closed. app/server.go:552
904	2021-11-05T18:00:27Z WARN [APP:SERVI] Failed to complete stream. error:[
905	ERROR REPORT:
906	Original Error: *trace.ConnectionProblemError context has cancelled before complete could succeed
907	Stack Trace:
908		/go/src/github.com/gravitational/teleport/lib/events/stream.go:403 github.com/gravitational/teleport/lib/events.(*ProtoStream).Complete
909		/go/src/github.com/gravitational/teleport/lib/events/emitter.go:471 github.com/gravitational/teleport/lib/events.(*CheckingStream).Complete
910		/go/src/github.com/gravitational/teleport/lib/events/auditwriter.go:478 github.com/gravitational/teleport/lib/events.(*AuditWriter).completeStream
911		/go/src/github.com/gravitational/teleport/lib/events/auditwriter.go:441 github.com/gravitational/teleport/lib/events.(*AuditWriter).processEvents
912		/go/src/github.com/gravitational/teleport/lib/events/auditwriter.go:65 github.com/gravitational/teleport/lib/events.NewAuditWriter.func1
913		/opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
914	User Message: context has cancelled before complete could succeed] events/auditwriter.go:479
915	    testing.go:967: TempDir RemoveAll cleanup: unlinkat /tmp/TestHandleConnection2724309182/001/log/upload/streaming/default: directory not empty
916	--- FAIL: TestHandleConnection (1.50s)
917	===================================================
918	exit status 1
919	Makefile:455: recipe for target 'test-go' failed
```

--- a/lib/events/auditwriter.go
+++ b/lib/events/auditwriter.go
@@ -473,8 +473,8 @@ func (a *AuditWriter) closeStream(stream apievents.Stream) {
 }
 
 func (a *AuditWriter) completeStream(stream apievents.Stream) {
-	// Cannot use the configured context b/c it's the servers and when the server
-	// is requested to close (and hence the context is cancled), the stream will not be able
+	// Cannot use the configured context because it's the server's and when the server
+	// is requested to close (and hence the context is canceled), the stream will not be able
 	// to complete
 	ctx, cancel := context.WithTimeout(context.Background(), defaults.NetworkBackoffDuration)
 	defer cancel()

--- a/lib/events/auditwriter.go
+++ b/lib/events/auditwriter.go
@@ -473,7 +473,10 @@ func (a *AuditWriter) closeStream(stream apievents.Stream) {
 }
 
 func (a *AuditWriter) completeStream(stream apievents.Stream) {
-	ctx, cancel := context.WithTimeout(a.cfg.Context, defaults.NetworkBackoffDuration)
+	// Cannot use the configured context b/c it's the servers and when the server
+	// is requested to close (and hence the context is cancled), the stream will not be able
+	// to complete
+	ctx, cancel := context.WithTimeout(context.Background(), defaults.NetworkBackoffDuration)
 	defer cancel()
 	if err := stream.Complete(ctx); err != nil {
 		a.log.WithError(err).Warning("Failed to complete stream.")

--- a/lib/events/filesessions/filestream.go
+++ b/lib/events/filesessions/filestream.go
@@ -142,7 +142,7 @@ func (h *Handler) CompleteUpload(ctx context.Context, upload events.StreamUpload
 		partPath := h.partPath(upload, part.Number)
 		file, err := os.Open(partPath)
 		if err != nil {
-			return trace.ConvertSystemError(err)
+			return trace.Wrap(err, "failed to open part file for upload")
 		}
 		files = append(files, file)
 		readers = append(readers, file)

--- a/lib/events/stream.go
+++ b/lib/events/stream.go
@@ -262,6 +262,7 @@ func NewProtoStream(cfg ProtoStreamConfig) (*ProtoStream, error) {
 		complete:     complete,
 		completeType: atomic.NewUint32(completeTypeComplete),
 		completeMtx:  &sync.RWMutex{},
+		writeDoneCh:  make(chan struct{}),
 
 		uploadsCtx:  uploadsCtx,
 		uploadsDone: uploadsDone,
@@ -318,6 +319,8 @@ type ProtoStream struct {
 	completeType   *atomic.Uint32
 	completeResult error
 	completeMtx    *sync.RWMutex
+	// writeDoneCh is closed when all streams have completed
+	writeDoneCh chan struct{}
 
 	// uploadsCtx is used to signal that all uploads have been completed
 	uploadsCtx context.Context
@@ -392,6 +395,7 @@ func (s *ProtoStream) EmitAuditEvent(ctx context.Context, event apievents.AuditE
 // Complete completes the upload, waits for completion and returns all allocated resources.
 func (s *ProtoStream) Complete(ctx context.Context) error {
 	s.complete()
+	<-s.writeDoneCh
 	select {
 	// wait for all in-flight uploads to complete and stream to be completed
 	case <-s.uploadsCtx.Done():
@@ -466,6 +470,7 @@ func (w *sliceWriter) trySendStreamStatusUpdate(lastEventIndex int64) {
 
 // receiveAndUpload receives and uploads serialized events
 func (w *sliceWriter) receiveAndUpload() {
+	defer close(w.proto.writeDoneCh)
 	// on the start, send stream status with the upload ID and negative
 	// index so that remote party can get an upload ID
 	w.trySendStreamStatusUpdate(-1)

--- a/lib/events/stream.go
+++ b/lib/events/stream.go
@@ -250,7 +250,6 @@ func NewProtoStream(cfg ProtoStreamConfig) (*ProtoStream, error) {
 	}
 	cancelCtx, cancel := context.WithCancel(context.Background())
 	completeCtx, complete := context.WithCancel(context.Background())
-	uploadsCtx, uploadsDone := context.WithCancel(context.Background())
 	stream := &ProtoStream{
 		cfg:      cfg,
 		eventsCh: make(chan protoEvent),
@@ -258,14 +257,11 @@ func NewProtoStream(cfg ProtoStreamConfig) (*ProtoStream, error) {
 		cancelCtx: cancelCtx,
 		cancel:    cancel,
 
-		completeCtx:  completeCtx,
-		complete:     complete,
-		completeType: atomic.NewUint32(completeTypeComplete),
-		completeMtx:  &sync.RWMutex{},
-		writeDoneCh:  make(chan struct{}),
-
-		uploadsCtx:  uploadsCtx,
-		uploadsDone: uploadsDone,
+		completeCtx:      completeCtx,
+		complete:         complete,
+		completeType:     atomic.NewUint32(completeTypeComplete),
+		completeMtx:      &sync.RWMutex{},
+		uploadLoopDoneCh: make(chan struct{}),
 
 		// Buffered channel gives consumers
 		// a chance to get an early status update.
@@ -319,13 +315,10 @@ type ProtoStream struct {
 	completeType   *atomic.Uint32
 	completeResult error
 	completeMtx    *sync.RWMutex
-	// writeDoneCh is closed when all streams have completed
-	writeDoneCh chan struct{}
 
-	// uploadsCtx is used to signal that all uploads have been completed
-	uploadsCtx context.Context
-	// uploadsDone is a function signalling that uploads have completed
-	uploadsDone context.CancelFunc
+	// uploadLoopDoneCh is closed when the slice exits the upload loop.
+	// The exit might be an indication of completion or a cancelation
+	uploadLoopDoneCh chan struct{}
 
 	// statusCh sends updates on the stream status
 	statusCh chan apievents.StreamStatus
@@ -395,14 +388,10 @@ func (s *ProtoStream) EmitAuditEvent(ctx context.Context, event apievents.AuditE
 // Complete completes the upload, waits for completion and returns all allocated resources.
 func (s *ProtoStream) Complete(ctx context.Context) error {
 	s.complete()
-	<-s.writeDoneCh
 	select {
-	// wait for all in-flight uploads to complete and stream to be completed
-	case <-s.uploadsCtx.Done():
+	case <-s.uploadLoopDoneCh:
 		s.cancel()
 		return s.getCompleteResult()
-	case <-s.cancelCtx.Done():
-		return trace.ConnectionProblem(s.cancelCtx.Err(), "emitter has been closed")
 	case <-ctx.Done():
 		return trace.ConnectionProblem(ctx.Err(), "context has cancelled before complete could succeed")
 	}
@@ -420,11 +409,8 @@ func (s *ProtoStream) Close(ctx context.Context) error {
 	s.completeType.Store(completeTypeFlush)
 	s.complete()
 	select {
-	// wait for all in-flight uploads to complete and stream to be completed
-	case <-s.uploadsCtx.Done():
-		return nil
-	case <-s.cancelCtx.Done():
-		return trace.ConnectionProblem(s.cancelCtx.Err(), "emitter has been closed")
+	case <-s.uploadLoopDoneCh:
+		return ctx.Err()
 	case <-ctx.Done():
 		return trace.ConnectionProblem(ctx.Err(), "context has cancelled before complete could succeed")
 	}
@@ -470,7 +456,7 @@ func (w *sliceWriter) trySendStreamStatusUpdate(lastEventIndex int64) {
 
 // receiveAndUpload receives and uploads serialized events
 func (w *sliceWriter) receiveAndUpload() {
-	defer close(w.proto.writeDoneCh)
+	defer close(w.proto.uploadLoopDoneCh)
 	// on the start, send stream status with the upload ID and negative
 	// index so that remote party can get an upload ID
 	w.trySendStreamStatusUpdate(-1)
@@ -499,7 +485,7 @@ func (w *sliceWriter) receiveAndUpload() {
 					return
 				}
 			}
-			defer w.completeStream()
+			w.completeStream()
 			return
 		case upload := <-w.completedUploadsC:
 			part, err := upload.getPart()
@@ -608,7 +594,6 @@ func (w *sliceWriter) submitEvent(event protoEvent) error {
 // completeStream waits for in-flight uploads to finish
 // and completes the stream
 func (w *sliceWriter) completeStream() {
-	defer w.proto.uploadsDone()
 	for range w.activeUploads {
 		select {
 		case upload := <-w.completedUploadsC:

